rs/athm/src/lib.rs

```diff
@@ -0,0 +1,1678 @@
+//! Anonymous Tokens with Hidden Metadata (ATHM)
+//!
+//! Privacy-preserving tokens where servers can embed metadata invisible to clients.
+//!
+//! # Overview
+//!
+//! ATHM allows servers to embed categorical metadata (like risk scores or user tiers)
+//! in tokens without clients knowing which metadata values. This enables privacy-preserving authentication
+//! with hidden authorization levels.
+//!
+//! # Example
+//!
+//! ```
+//! use athm::*;
+//! use rand_core::OsRng;
+//!
+//! // Setup with 4 metadata buckets (e.g., risk levels 0-3)
+//! let params = Params::new(4).unwrap();
+//! let (private_key, public_key, proof) = key_gen(&params);
+//!
+//! // Client creates blinded request
+//! let mut rng = OsRng;
+//! let (context, request) = token_request(&public_key, &proof, &params, &mut rng).unwrap();
+//!
+//! // Server responds with hidden metadata
+//! let hidden_metadata = 2;
+//! let response = token_response(
+//!     &private_key, &public_key, &request, hidden_metadata, &params, &mut rng
+//! ).unwrap();
+//!
+//! // Client unblinds token
+//! let token = finalize_token(
+//!     &context, &public_key, &request, &response, &params, &mut rng
+//! ).unwrap();
+//!
+//! // Server verifies and recovers metadata
+//! let metadata = verify_token(&private_key, &token, &params).unwrap();
+//! assert_eq!(metadata, hidden_metadata);
+//! ```
+//!
+//! # Protocol Flow
+//!
+//! 1. Server generates keys with [`key_gen`]
+//! 2. Client creates blinded request with [`token_request`]
+//! 3. Server embeds metadata with [`token_response`]
+//! 4. Client unblinds with [`finalize_token`]
+//! 5. Server verifies with [`verify_token`]
+
+use elliptic_curve::generic_array::typenum::Unsigned;
+use elliptic_curve::hash2curve::{ExpandMsgXmd, GroupDigest};
+use p256::{
+    elliptic_curve::{
+        bigint::U256,
+        group::GroupEncoding,
+        ops::Reduce,
+        sec1::{ModulusSize, ToEncodedPoint},
+        subtle::{Choice, ConditionallySelectable, ConstantTimeEq, CtOption},
+        Field, FieldBytes, FieldBytesSize, Group, PrimeField,
+    },
+    NistP256, NonZeroScalar, ProjectivePoint, Scalar,
+};
+use rand_core::{CryptoRngCore, OsRng};
+use sha2::{Digest, Sha256};
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+/// A transcript for Fiat-Shamir transform
+///
+/// This struct provides a clean abstraction for building transcripts
+/// and generating challenges in non-interactive zero-knowledge proofs.
+///
+/// The transcript follows a similar pattern to anonymous-credit-tokens:
+/// - Elements are added with labels for clarity
+/// - Points and scalars are length-prefixed
+/// - Domain separation is applied when generating challenges
+#[derive(Clone)]
+struct Transcript {
+    hasher: Sha256,
+}
+
+// TODO(samschlesinger) add info string to spec and here for cleaner domain separation
```

Comment:
Typically, we scrub TODOs from github repos.

Reply:
Done.